app/assets/javascripts/reducers/revisions.js

```diff
@@ -76,7 +76,15 @@ export default function revisions(state = initialState, action) {
     case SORT_REVISIONS: {
       const absolute = action.key === 'characters';
       const desc = action.key === state.sort.sortKey;
-      const sortedRevisions = sortByKey(state.revisions, action.key, null, desc, absolute);
+      // some values like references and rating num are fetched independently of revisions
+      // to reduce waiting time
+      let mapTo;
+      if (action.key === 'references_added') {
```

Comment:
I don't have a clear picture of why the key function is necessary. What is different about this situation that we can't just pass the key to the sort function like with all the other sorts that happen currently?

Reply:
Since the page assessments and references added are fetched after the revisions have been rendered, they exist as different redux states. Before this, all the properties were part of the revision object - now, that's not the case. The references added state for example is an object with a key of the revision id and the value of how many references were added.

Reply:
Ah, I see. Why not add them to the `revision` objects in redux whenever they get received?

Reply:
I didn't want to mutate the revisions array since that could lead to more problems. Say if the user fetches revisions and they're loaded in. If page assessments comes in first,  sets `state.revisions.revisions` to this a new array(which has the class info), then this change would be overridden by the references request, which would do the same - set `state.revision.revisions` to the its version of the array. Not to mention that if the user presses see more before either of them can come in, they will have a different version of the revisions array. 

I just felt a lot of things could go wrong, although I assume most of the times they won't. This way it doesn't matter at all in which order the requests are completed since all 3 requests would work on 3 different redux states.

Reply:
Can't it be done in a way that doesn't risk any data being overwritten? The assessments received redux action should only add the new data to the relevant revisions, not replace or remove any other attributes.

Like, first you fetch revisions and have an array like `[{ id: 123, title: 'abc' }]`, and then you receive page assessment data and it becomes `[{ id: 123, title: 'abc', class: 'c' }]`, and then you receive ORES data and it becomes `[{ id: 123, title: 'abc', class: 'c', references_added: 2 }]`. It should be possible to write reducer functions such that the order of receiving data doesn't matter.

Reply:
This kind of situation is where redux really shows its value, in my opinion. We do something similar in the `article_finder.js` reducer.